bluefog/common/operations.cc

```diff
@@ -186,151 +504,401 @@ bool RunLoopOnce(BluefogGlobalState& state) {
+          bluefog_global.nccl_controller->WinPut(entry);
         }
 #endif
-        state.timeline.ActivityEnd(entry.tensor_name);
+        timeline.ActivityEnd(entry.tensor_name);
         break;
       case MPIOpsType::WIN_GET:
         BFLOG(TRACE, bluefog_global.controller->GetRank())
             << "Processing WIN_GET on " << entry.tensor_name << " with "
             << Vendor_Name(controller_vendor);
-        state.timeline.ActivityStart(entry.tensor_name, "WIN_GET");
+        timeline.ActivityStart(entry.tensor_name, "WIN_GET");
         if (controller_vendor == Vendor::MPI) {
-          state.controller->WinGet(entry);
+          bluefog_global.controller->WinGet(entry);
         }
 #if HAVE_NCCL
         if (controller_vendor == Vendor::NCCL) {
-          state.nccl_controller->WinGet(entry);
+          bluefog_global.nccl_controller->WinGet(entry);
         }
 #endif
-        state.timeline.ActivityEnd(entry.tensor_name);
+        timeline.ActivityEnd(entry.tensor_name);
         break;
       case MPIOpsType::WIN_ACCUMULATE:
         BFLOG(TRACE, bluefog_global.controller->GetRank())
             << "Processing WIN_ACCUMULATE on " << entry.tensor_name << " with "
             << Vendor_Name(controller_vendor);
-        state.timeline.ActivityStart(entry.tensor_name, "WIN_ACCUMULATE");
+        timeline.ActivityStart(entry.tensor_name, "WIN_ACCUMULATE");
         if (controller_vendor == Vendor::MPI) {
-          state.controller->WinAccumulate(entry);
+          bluefog_global.controller->WinAccumulate(entry);
         }
 #if HAVE_NCCL
         if (controller_vendor == Vendor::NCCL) {
-          state.nccl_controller->WinAccumulate(entry);
+          bluefog_global.nccl_controller->WinAccumulate(entry);
+        }
+#endif
+        timeline.ActivityEnd(entry.tensor_name);
+        break;
+      case MPIOpsType::WIN_CREATE:
+        BFLOG(TRACE, bluefog_global.controller->GetRank())
+            << "Processing WIN_CREATE " << entry.tensor_name << " with "
+            << Vendor_Name(controller_vendor);
+#if HAVE_NCCL
+        if (controller_vendor == Vendor::NCCL) {
+          if (!nccl_context.is_initialized) {
+            bluefog_global.nccl_controller->Initialize();
+            BFLOG(INFO, bluefog_global.controller->GetRank())
+                << "NCCL Initialized";
+          }
+          bluefog_global.nccl_controller->WinCreate(entry);
         }
 #endif
-        state.timeline.ActivityEnd(entry.tensor_name);
+        if (controller_vendor == Vendor::MPI) {
+          bluefog_global.controller->WinCreate(entry);
+        }
+        break;
+      case MPIOpsType::WIN_FREE:
+        BFLOG(TRACE, bluefog_global.controller->GetRank())
+            << "Processing WIN_FREE " << entry.tensor_name << " with "
+            << Vendor_Name(controller_vendor);
+#if HAVE_NCCL
+        // No specified name. So both nccl and mpi will Free win.
+        if (nccl_context.is_initialized && entry.tensor_name.empty()) {
```

Comment:
It is because the controller_vendor is dependent on the device of the stored tensor. However, the empty name indicates all windows will be freed, which cannot determine the device.